homeassistant/components/arcam_fmj/__init__.py

```diff
@@ -0,0 +1,160 @@
+    DEFAULT_PORT,
+    DEFAULT_SCAN_INTERVAL,
+    SIGNAL_CLIENT_DATA,
+    SIGNAL_CLIENT_STARTED,
+    SIGNAL_CLIENT_STOPPED,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _optional_zone(value):
+    if value:
+        return ZONE_SCHEMA(value)
+    return ZONE_SCHEMA({})
+
+
+ZONE_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Optional(SERVICE_TURN_ON): cv.SERVICE_SCHEMA,
+    }
+)
+
+DEVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PORT, default=DEFAULT_PORT): cv.positive_int,
+        vol.Optional(CONF_ZONE): {vol.In([1, 2]): _optional_zone},
+        vol.Optional(
+            CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
+        ): cv.positive_int,
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.All(cv.ensure_list, [DEVICE_SCHEMA])}, extra=vol.ALLOW_EXTRA
+)
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType):
+    """Set up the component."""
+    hass.data[DOMAIN_DATA_ENTRIES] = {}
+    hass.data[DOMAIN_DATA_CONFIG] = {}
+
+    for device in config[DOMAIN]:
+        hass.data[DOMAIN_DATA_CONFIG][
+            (device[CONF_HOST], device[CONF_PORT])
+        ] = device
+
+        hass.async_add_job(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": config_entries.SOURCE_IMPORT},
+                data={
+                    CONF_HOST: device[CONF_HOST],
+                    CONF_PORT: device[CONF_PORT],
+                },
+            )
+        )
+
+    return True
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType, entry: config_entries.ConfigEntry
+):
+    """Set up an access point from a config entry."""
+    client = Client(entry.data[CONF_HOST], entry.data[CONF_PORT])
+
+    config = hass.data[DOMAIN_DATA_CONFIG].get(
+        (entry.data[CONF_HOST], entry.data[CONF_PORT]),
+        DEVICE_SCHEMA(
+            {
+                CONF_HOST: entry.data[CONF_HOST],
+                CONF_PORT: entry.data[CONF_PORT],
+            }
+        ),
+    )
+
+    hass.data[DOMAIN_DATA_ENTRIES][entry.entry_id] = {
+        "client": client,
+        "config": config,
+    }
+
+    asyncio.ensure_future(
+        _run_client(hass, client, config[CONF_SCAN_INTERVAL])
+    )
+
+    hass.async_add_job(
+        hass.config_entries.async_forward_entry_setup(entry, "media_player")
+    )
+
+    return True
+
+
+async def _run_client(hass, client, interval):
+    task = asyncio.Task.current_task()
+    run = True
+
+    async def _stop(_):
+        nonlocal run
+        run = False
+        task.cancel()
+        await task
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _stop)
+
+    def _listen(_):
+        hass.helpers.dispatcher.async_dispatcher_send(
+            SIGNAL_CLIENT_DATA, client.host
+        )
+
+    while run:
+        try:
+            await asyncio.wait_for(client.start(), timeout=interval)
```

Comment:
We can use `async_timeout`:
```suggestion
            with async_timeout.timeout(interval):
                await client.start()
```